The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27349446203/job/80807071427) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/80807071427) [(plain)](https://github.com/rust-lang/rust/commit/d3a02b06aca872710761206c43cca128d54184bd/checks/80807071427/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/layout/issue-unsized-tail-restatic-ice-122488.rs stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/layout/issue-unsized-tail-restatic-ice-122488/issue-unsized-tail-restatic-ice-122488.stderr`
diff of stderr:

22 LL | struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(Box<V>, U);
23    |                                                             ++++ +
24 
- error[E0080]: the type `ArenaSet<Vec<u8>, [u8]>` has an unknown layout
-   --> $DIR/issue-unsized-tail-restatic-ice-122488.rs:8:1
-    |
- LL | const DATA: *const ArenaSet<Vec<u8>> = std::ptr::null_mut();
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `DATA` failed here
+ error: aborting due to 1 previous error
30 
- error: aborting due to 2 previous errors
- 
- Some errors have detailed explanations: E0080, E0277.
---
To only update this specific test, also pass `--test-args layout/issue-unsized-tail-restatic-ice-122488.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/layout/issue-unsized-tail-restatic-ice-122488.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/layout/issue-unsized-tail-restatic-ice-122488" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0277]: the size for values of type `V` cannot be known at compilation time
##[error]  --> /checkout/tests/ui/layout/issue-unsized-tail-restatic-ice-122488.rs:5:61
   |
LL | struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(V, U);
   |                           --------------------------------  ^ doesn't have a size known at compile-time
   |                           |
   |                           this type parameter needs to be `Sized`
   |
   = note: only the last field of a struct may have a dynamically sized type
   = help: change the field's type to have a statically known size
help: consider removing the `?Sized` bound to make the type parameter `Sized`
   |
LL - struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(V, U);
LL + struct ArenaSet<U: Deref, V = <U as Deref>::Target>(V, U);
   |
help: borrowed types always have a statically known size
   |
LL | struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(&V, U);
   |                                                             +
help: the `Box` type always has a statically known size and allocates its contents in the heap
   |
LL | struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(Box<V>, U);
   |                                                             ++++ +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
------------------------------------------

---- [ui] tests/ui/layout/issue-unsized-tail-restatic-ice-122488.rs stdout end ----
---- [ui] tests/ui/sized/stack-overflow-trait-infer-98842.rs stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sized/stack-overflow-trait-infer-98842/stack-overflow-trait-infer-98842.stderr`
diff of stderr:

7 note: ...which requires computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`...
8   --> $SRC_DIR/core/src/ops/deref.rs:LL:COL
9    = note: ...which again requires computing layout of `Foo`, completing the cycle
- note: cycle used when const-evaluating + checking `_`
-   --> $DIR/stack-overflow-trait-infer-98842.rs:13:1
-    |
- LL | const _: *const Foo = 0 as _;
-    | ^^^^^^^^^^^^^^^^^^^
+ note: cycle used when computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`
+   --> $SRC_DIR/core/src/ops/deref.rs:LL:COL
15    = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
16 
17 error: aborting due to 1 previous error

Note: some mismatched output was normalized before being compared
-   --> /rustc/FAKE_PREFIX/library/core/src/ops/deref.rs:144:4
+ note: cycle used when computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`
+   --> $SRC_DIR/core/src/ops/deref.rs:LL:COL


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args sized/stack-overflow-trait-infer-98842.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sized/stack-overflow-trait-infer-98842.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sized/stack-overflow-trait-infer-98842" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021"
stdout: none
--- stderr -------------------------------
error[E0391]: cycle detected when computing layout of `Foo`
##[error]  --> /checkout/tests/ui/sized/stack-overflow-trait-infer-98842.rs:8:1
   |
LL | struct Foo(<&'static Foo as ::core::ops::Deref>::Target);
   | ^^^^^^^^^^
   |
note: ...which requires computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`...
  --> /rustc/FAKE_PREFIX/library/core/src/ops/deref.rs:144:4
   = note: ...which again requires computing layout of `Foo`, completing the cycle
note: cycle used when computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`
  --> /rustc/FAKE_PREFIX/library/core/src/ops/deref.rs:144:4
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

error: aborting due to 1 previous error

```

</details>

--- a/compiler/rustc_const_eval/src/interpret/validity.rs
+++ b/compiler/rustc_const_eval/src/interpret/validity.rs
@@ -925,7 +925,7 @@ impl<'rt, 'tcx, M: Machine<'tcx>> ValidityVisitor<'rt, 'tcx, M> {
                 }
                 interp_ok(true)
             }
-            ty::RawPtr(..) => {
+            ty::RawPtr(pointee, ..) => {
                 let ptr = self.read_immediate(value, ExpectedKind::RawPtr)?;
                 if self.reset_provenance_and_padding {
                     self.reset_pointer_provenance(value, &ptr)?;
@@ -933,8 +933,12 @@ impl<'rt, 'tcx, M: Machine<'tcx>> ValidityVisitor<'rt, 'tcx, M> {
                     self.add_data_range_place(value);
                 }
 
-                let place = self.ecx.imm_ptr_to_mplace(&ptr)?;
-                if place.layout.is_unsized() {
+                if !pointee.is_sized(*self.ecx.tcx, self.ecx.typing_env) {
+                    // Raw pointers to unsized types need to have their metadata checked.
+                    // We avoid creating this place for sized types to match codegen: those types
+                    // might actually be invalid (i.e., too big)!
+                    let place = self.ecx.imm_ptr_to_mplace(&ptr)?;
+                    assert!(place.layout.is_unsized());
                     self.check_wide_ptr_meta(place.meta(), place.layout)?;
                 }
                 interp_ok(true)

--- a/src/tools/miri/tests/pass/too-big-type-behind-raw-ptr.rs
+++ b/src/tools/miri/tests/pass/too-big-type-behind-raw-ptr.rs
@@ -1,0 +1,10 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/157654#issuecomment-4679655886>:
+//! the type behind a raw pointer should never have its layout computed.
+
+//@compile-flags: -Zmiri-permissive-provenance
+
+const PTR_BITS_MINUS_1: usize = std::mem::size_of::<*const ()>() * 8 - 1;
+
+fn main() {
+    std::hint::black_box(0 as *const [u64; 1 << PTR_BITS_MINUS_1]);
+}

--- a/tests/ui/layout/issue-unsized-tail-restatic-ice-122488.rs
+++ b/tests/ui/layout/issue-unsized-tail-restatic-ice-122488.rs
@@ -5,7 +5,7 @@ use std::ops::Deref;
 struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(V, U);
 //~^ ERROR the size for values of type `V` cannot be known at compilation time
 
-const DATA: *const ArenaSet<Vec<u8>> = std::ptr::null_mut();
+const DATA: &ArenaSet<Vec<u8>> = unsafe { &* std::ptr::null_mut() };
 //~^ ERROR the type `ArenaSet<Vec<u8>, [u8]>` has an unknown layout
 
 pub fn main() {}

--- a/tests/ui/layout/issue-unsized-tail-restatic-ice-122488.stderr
+++ b/tests/ui/layout/issue-unsized-tail-restatic-ice-122488.stderr
@@ -23,10 +23,10 @@ LL | struct ArenaSet<U: Deref, V: ?Sized = <U as Deref>::Target>(Box<V>, U);
    |                                                             ++++ +
 
 error[E0080]: the type `ArenaSet<Vec<u8>, [u8]>` has an unknown layout
-  --> $DIR/issue-unsized-tail-restatic-ice-122488.rs:8:1
+  --> $DIR/issue-unsized-tail-restatic-ice-122488.rs:8:43
    |
-LL | const DATA: *const ArenaSet<Vec<u8>> = std::ptr::null_mut();
-   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `DATA` failed here
+LL | const DATA: &ArenaSet<Vec<u8>> = unsafe { &* std::ptr::null_mut() };
+   |                                           ^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `DATA` failed here
 
 error: aborting due to 2 previous errors
 

--- a/tests/ui/sized/stack-overflow-trait-infer-98842.stderr
+++ b/tests/ui/sized/stack-overflow-trait-infer-98842.stderr
@@ -7,11 +7,8 @@ LL | struct Foo(<&'static Foo as ::core::ops::Deref>::Target);
 note: ...which requires computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`...
   --> $SRC_DIR/core/src/ops/deref.rs:LL:COL
    = note: ...which again requires computing layout of `Foo`, completing the cycle
-note: cycle used when const-evaluating + checking `_`
-  --> $DIR/stack-overflow-trait-infer-98842.rs:13:1
-   |
-LL | const _: *const Foo = 0 as _;
-   | ^^^^^^^^^^^^^^^^^^^
+note: cycle used when computing layout of `<&'static Foo as core::ops::deref::Deref>::Target`
+  --> $SRC_DIR/core/src/ops/deref.rs:LL:COL
    = note: for more information, see <https://rustc-dev-guide.rust-lang.org/overview.html#queries> and <https://rustc-dev-guide.rust-lang.org/query.html>
 
 error: aborting due to 1 previous error

